### PR TITLE
[3752] [WIP] Filter support for invoice_payload

### LINKED ISSUE #3752: Filter support for invoice_payload in PreCheckoutQueryHandler and successful_payment message
### What kind of feature are you missing? Where do you notice a shortcoming of PTB?

There is currently no solution to filter for `invoice_payload` in the previously mentioned updates.

### Describe the solution you'd like

I would like to change the SuccessfulPayment message filter to the same type of filter we have with the [text one](https://docs.python-telegram-bot.org/telegram.ext.filters.html#telegram.ext.filters.Text). And I would like to have a pattern argument in the PreCheckoutQueryHandler, similar to the [callbackqueryhandler](https://docs.python-telegram-bot.org/en/stable/telegram.ext.callbackqueryhandler.html).

### Describe alternatives you've considered

Filtering has to happen in the callback. That works, but is not our design philosophy.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by erickisos**
Closes #3752

### CODE REVIEW

**Comment by Bibo-Joshi** [comment]
Thanks for the PR! Kindly leave a short comment once you consider your PR ready for review :)

File: telegram/ext/filters.py [hunk 1]
```diff
@@ -958,30 +994,42 @@ def filter(self, message: Message) -> bool:
...
+        self.values: Optional[Collection[int]] = (
+            [values] if isinstance(values, int) else values
+        )
 
         if emoji:  # for filters.Dice.BASKETBALL
             self.name = f"filters.Dice.{emoji.name}"
-            if self.values and emoji:  # for filters.Dice.Dice(4)  SLOT_MACHINE -> SlotMachine
+            if (
+                self.values and emoji
+            ):  # for filters.Dice.Dice(4)  SLOT_MACHINE -> SlotMachine
```
Comment (harshil21) [inline telegram/ext/filters.py:1012]: can you revert such formatting changes? They are not required and are cluttering this PR.

File: telegram/ext/filters.py [hunk 2]
```diff
@@ -958,30 +994,42 @@ def filter(self, message: Message) -> bool:
...
 
     def filter(self, message: Message) -> bool:
-        if not (dice := message.dice):  # no dice
+        if not message.dice:  # no dice
             return False
 
         if self.emoji:
-            emoji_match = dice.emoji == self.emoji
+            emoji_match = message.dice.emoji == self.emoji
             if self.values:
```
Comment (harshil21) [inline telegram/ext/filters.py:1025]: please revert all these changes

File: telegram/ext/filters.py [hunk 3]
```diff
@@ -2144,28 +2245,30 @@ def filter(self, message: Message) -> bool:
...
-    def filter(self, message: Message) -> bool:
-        return bool(message.story)
-
-
-STORY = _Story(name="filters.STORY")
-"""Messages that contain :attr:`telegram.Message.story`.
-
-.. versionadded:: 20.5
-"""
-
```
Comment (harshil21) [inline telegram/ext/filters.py:2159]: please add this back

**Review by harshil21 — CHANGES_REQUESTED** [review]
thanks for the PR, I didn't review the core part yet since there seems to be lots of unwanted changes. Can you fix those first (maybe its better to start over)?

File: telegram/ext/filters.py (hunk 1 above)
Comment (erickisos) [reply telegram/ext/filters.py:1012]: Let me take a look on those changes then, thanks for the review

**Comment by Bibo-Joshi** [comment]
closing due to inactivity

**Pull request closed without merging**